src/librustc_trans/partitioning.rs

```diff
@@ -410,3 +501,76 @@ fn compute_codegen_unit_name<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
     return token::intern_and_get_ident(&mod_path[..]);
 }
+
+fn single_codegen_unit<'a, 'tcx, I>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                    trans_items: I,
+                                    reachable: &NodeSet)
+                                    -> CodegenUnit<'tcx>
+    where I: Iterator<Item = TransItem<'tcx>>
+{
+    let mut items = FnvHashMap();
+
+    for trans_item in trans_items {
+        let linkage = trans_item.explicit_linkage(tcx).unwrap_or_else(|| {
+            match trans_item {
+                TransItem::Static(node_id) => {
+                    if reachable.contains(&node_id) {
+                        llvm::ExternalLinkage
+                    } else {
+                        llvm::PrivateLinkage
+                    }
+                }
+                TransItem::DropGlue(_) => {
+                    llvm::LinkOnceODRLinkage
+                }
+                TransItem::Fn(instance) => {
+                    if trans_item.is_generic_fn() {
+                        llvm::WeakODRLinkage
+                    } else if trans_item.is_from_extern_crate() {
+                        llvm::AvailableExternallyLinkage
+                    } else if reachable.contains(&tcx.map
+                                                     .as_local_node_id(instance.def)
+                                                     .unwrap()) {
+                        llvm::ExternalLinkage
+                    } else {
+                        llvm::InternalLinkage
+                    }
```

Comment:
The reasoning behind the externally available one is that non-generic, inlined functions must already be instantiated in the upstream crate. Since they should also always be in the `reachable` set, they won't be made internal later on either.

All monomorphizations should already be using weak_odr (when multiple codegen units are activated) today. 

Also, as far as I could tell, all the undefined symbols in the error where for functions from the local crate.

Thank you so much for helping look into this, by the way!